addons/bus/websocket.py

```diff
@@ -641,16 +675,39 @@ def _dispatch_bus_notifications(self):
         session = root.session_store.get(self._session.sid)
         if not session:
             raise SessionExpiredException()
+        # Discard all the smallest notification ids that have expired and
+        # increment the last id accordingly. History can only be trimmed of ids
+        # that are below the new last id otherwise some notifications might be
+        # dispatched again.
+        # For example, if the theshold is 10s, and the state is:
+        # last id 2, history [(3, 8s), (6, 10s), (7, 7s)]
+        # If 6 is removed because it is above the threshold, the next query will
+        # be (id > 2 AND id NOT IN (3, 7)) which will fetch 6 again.
+        # 6 can only be removed after 3 reaches the threshold and is removed as
+        # well, and if 4 appears in the meantime, 3 can be removed but 6 will
+        # have to wait for 4 to reach the threshold as well.
+        last_index = -1
+        for i, notif in enumerate(self._notif_history):
+            if time.time() - notif[1] > self.MAX_NOTIFICATION_HISTORY_SEC:
+                last_index = i
+            else:
+                break
+        if last_index != -1:
+            self._last_notif_sent_id = self._notif_history[last_index][0]
+            self._notif_history = self._notif_history[last_index + 1 :]
         with acquire_cursor(session.db) as cr:
             env = api.Environment(cr, session.uid, dict(session.context, lang=None))
             if session.uid is not None and not check_session(session, env):
                 raise SessionExpiredException()
             # Mark the notification request as processed.
             self.__notif_sock_r.recv(1)
-            notifications = env['bus.bus']._poll(self._channels, self._last_notif_sent_id)
+            notifications = env["bus.bus"]._poll(
+                self._channels, self._last_notif_sent_id, [n[0] for n in self._notif_history]
+            )
```

Comment:
I think it could be possible for notifications to be sent twice:

- `_last_notif_sent_id = 0` initially
- notif with id `1` not yet commit
- notif with id `2` commit
- notif with id `3` commit
- notif with id `1` fully commit (after 10 seconds of `2`)

What could happen is `_notif_history` to look like this:
- `[]`
- `[2]`
- `[2, 3]`
- `[1, 3]` (`2` is 10 seconds old)

So `_last_notif_sent_id` stays at `0`, but in the last step since `2` is not in `_notif_history` but `_last_notif_sent_id` is still 0, it will pick `2` in `_poll()` and be resend again.

Reply:
Note that in your example it should be impossible to remove id 2 and keep `_last_notif_sent_id` at 0.
Whenever we remove values, the last id is incremented to this value (if we remove 2, last becomes 2, if we remove [4, 5, 6], last becomes 6).

Either we have last 0, [2, 3] (1 will be fetched) or last 2, [3] (1 will be missed).
In what situation do you think we could have 0, [3]? If that happens indeed we'll have a duplicate, but I don't think it's possible with the current code.

But for the rest of the flow, this is exactly how it works and it is the limitation of this PR, we give 10s for id 1 to commit after id 2 has been dispatched before we consider it obsolete and drop it. So it can still happen to miss notifs but the conditions are more controlled and should never happen in practice with normal business code.

As explained in the comment in the code, id 1 needs to commit faster or don't call "create bus" so far away from the end of its transaction to avoid the issue.
We can also increase the delay a little bit, but with the current architecture there will always be a moment where notifs are missed in that case, unless we save the full fetched list without ever using last id, which we don't want to do.